src/client/java/minicraft/core/Initializer.java

```diff
@@ -96,12 +96,23 @@ static void run() {
 			}
 
 			now = System.nanoTime();
-			if (now >= lastRender + 1E9D / MAX_FPS) {
+			if (now >= lastRender + 1E9D / MAX_FPS / 1.01) {
 				frames++;
 				lastRender = now;
 				Renderer.render();
 			}
 
+			try {
+				long curNano = System.nanoTime();
+				long untilNextTick = (long) (lastTick + nsPerTick - curNano);
+				long untilNextFrame = (long) (lastRender + 1E9D / MAX_FPS - curNano);
+				if (untilNextTick > 1E3 && untilNextFrame > 1E3) {
+					double timeToWait = Math.min(untilNextTick, untilNextFrame) / 1.2; // in nanosecond
+					//noinspection BusyWait
+					Thread.sleep((long) Math.floor(timeToWait / 1E6), (int) ((timeToWait - Math.floor(timeToWait)) % 1E6));
```

Comment:
You haven't really fixed the problem.
"In most cases spinning is considered an [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern) and should be avoided,[[2]](https://en.wikipedia.org/wiki/Busy_waiting#cite_note-2) as processor time that could be used to execute a different [task](https://en.wikipedia.org/wiki/Thread_(computer_science)) is instead wasted on useless activity."
- https://en.wikipedia.org/wiki/Busy_waiting

Reply:
> Busy-waiting itself can be made much less wasteful by using a delay function (e.g., [sleep()](https://en.wikipedia.org/wiki/Sleep())) found in most operating systems. This puts a thread to sleep for a specified time, during which the thread will waste no CPU time. If the loop is checking something simple then it will spend most of its time asleep and will waste very little CPU time.

Reply:
Your IDE begs to differ. The functions might not neccecaraly be the same.

Reply:
Why is it?

Reply:
I don't understand what you mean with "Why is it?" but, this is a line where you force the IDE to shut up.
//noinspection BusyWait

Reply:
IDE expects that using `wait()` in a loop is a worse practice because it may appears as using this function to pause the current thread *for another thread*. Here, this main thread only executes `tick()` and `render()` in desired intervals. Also, for constant interval, using timeout is better, but this does not. By the way, it would be better to separate rendering and ticking into 2 threads to use `waiting` in the better way, since it is currently waiting for both next ticking and rendering and wait for the shortest interval to execute the closest one and it would be *weird*, but concurrency and synchronization are required to be handled in an ideal way.

Reply:
Yeah, that sounds good.